### PR TITLE
feat: Configurable accounting dimension filters and validations

### PR DESCRIPTION
**Pull request by deepeshgarg007**
This feature allows user to apply restrictions on various accounting dimensions

Configuring Accounting Dimension Filters:
image

Filtering based on the accounting dimension filter:
image

Validations on GL Entry:
image

image

docs: https://github.com/frappe/erpnext_documentation/pull/195

### CODE REVIEW

File: erpnext/accounts/doctype/accounting_dimension/accounting_dimension.py [hunk 1]
```diff
@@ -203,7 +203,7 @@ def get_dimension_with_children(doctype, dimension):
 	return all_dimensions
 
 @frappe.whitelist()
-def get_dimension_filters():
+def get_dimension_filters(with_costcenter_and_project=False):
```
Comment (rmehta) [inline erpnext/accounts/doctype/accounting_dimension/accounting_dimension.py]: `cost_center`

File: erpnext/controllers/queries.py [hunk 2]
```diff
@@ -493,6 +493,53 @@ def get_income_account(doctype, txt, searchfield, start, page_len, filters):
...
+			query_selector = 'in'
+		else:
+			query_selector = 'not in'
+
+		if len(dimension_filters['allowed_dimensions']) == 1:
+			dimensions = tuple(dimension_filters['allowed_dimensions'] * 2)
+		else:
+			dimensions = tuple(dimension_filters['allowed_dimensions'])
+
+		result = frappe.db.sql("""SELECT name from `tab{doctype}` where
```
Comment (rmehta) [inline erpnext/controllers/queries.py]: use ORM

File: erpnext/public/js/utils/dimension_tree_filter.js [hunk 3]
```diff
@@ -1,18 +1,21 @@
 frappe.provide('frappe.ui.form');
-
 let default_dimensions = {};
```
Comment (rmehta) [inline erpnext/public/js/utils/dimension_tree_filter.js]: file needs to be namespaced

Comment (rmehta) [reply erpnext/public/js/utils/dimension_tree_filter.js]:
```
erpnext.accounts.dimensions = {
  setup(doctype) {
     // if (doctype) is applicable, then set the filters
     this.fetch_custom_dimensions(doctype).then(() => this.setup_filters());
  }

  fetch_custom_dimensions(doctype) {

  }
}
```

in `sales_invoice.js`

```
onload(frm) {
   erpnext.accounts.dimensions.setup(frm.doctype);
   ...
}
```

**Review by rmehta — CHANGES_REQUESTED** [review]

File: erpnext/accounts/doctype/accounting_dimension_filter/test_accounting_dimension_filter.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2020, Frappe Technologies Pvt. Ltd. and Contributors
+# See license.txt
+from __future__ import unicode_literals
+
+import frappe
+import unittest
+from erpnext.accounts.doctype.sales_invoice.test_sales_invoice import create_sales_invoice
+from erpnext.accounts.doctype.accounting_dimension.test_accounting_dimension import create_dimension, disable_dimension
+from erpnext.exceptions import InvalidAccountDimension, MandatoryDimension
```
Comment (netchampfaris) [inline erpnext/accounts/doctype/accounting_dimension_filter/test_accounting_dimension_filter.py]: Consistent error class naming
```suggestion
from erpnext.exceptions import InvalidAccountDimensionError, MandatoryAccountDimensionError
```

**Pull request merged**